WebDevStudios/wd_s
# Integrate Coding Standards Linting, Auto-formatting, & Husky.

See #548 

Integrate coding standards + husky so the code is so clean you could eat off of it!

### DESCRIPTION ###

- Beefed-up support for stylelint, _but did not get it ported to css-coding-standards, so using custom config here_
- Improvements to `package.json` scripts using `@webdevstudios/npm-scripts`<sup>([npmjs.org](https://github.com/WebDevStudios/npm-scripts))</sup>
- Reduced configuration files by porting most to `package.json`
- Supreme ignore patterns to control developer workflow no matter what extensions you have installed 🤞🏻
- Auto-formatting via `@webdevstudios/npm-scripts`<sup>([npmjs.org](https://github.com/WebDevStudios/npm-scripts))</sup> using `eslint --fix` and `stylelint --fix` _instead of Prettier_
- But, _Prettier added independently_ to format extraneous files, e.g. `.md`, `.json`, `.yml`, etc (we do not use for JS and SCSS)
- Adds something to `npm test` so we can do a quick test to make sure things are healthy
- Husky integrated, which auto-formats staged files and lints them, failing if commit does not meet standards after auto-formatting procedures
- Custom stylelint rules (not using our css-coding-standards, _yet_) and modifications to make compatible with linting and auto-formatting
- `npm test` does some health checks to ensure things are working properly as a package overall
- `preinstall` hook to ensure using `node@12` (warning only) <sup>[(Screenshot)](http://aubrey.pw/d/2020/1602866494.png)</sup>

### @gregrickaby on #548 

> Code editors should be able to find coding standard configs

All configurations are set for JavaScript, PHP, and Styles in `package.json` and various configurations, but editors should detect them automagically.

Tested in Sublime

- [ ] Needs VSCode confirmation

> Code editors should be able to auto-format on save (yes, this includes support for Prettier)

As discussed, we did _not_ go with Prettier for JS, but instead `eslint --fix` and `sass-lint --fix` for styles, we do however use Prettier for everything else, e.g. `.md`, `.json`, files etc.

> @wordpress/scripts should also be linting

It is, but with a layer of control called [@webdevstudios/npm-scripts](https://github.com/WebDevStudios/npm-scripts), which utilizes `wp-scripts` for all commands other than `format-js` and `format-styles`, because, as mentioned above, we do not want to use.

@webdevstudios/scripts also gives us a layer of control on future projects, ensuring when you run commands in `package.json`'s `.scripts` you just set it-and forget it™.

> Possibly bring in Husky and use Lint Staged to enforce coding standards on the pre-commit hook. (See Greg if you have questions)

Husky is integrated, and runs auto-formatting (per configs), and linting (per configs) either allowing or _not_ allowing a commit depending if code meets standards.

#### TODO

- [x] Get it linting and complaining
- [x] Get it auto-formatting your stuff
- [x] Make sure it won't let you put it into the universe, unless it's clean!

### ~SCREENSHOTS~ Video *click* ###

[![](https://img.youtube.com/vi/lg-KsjrlORo/0.jpg)](https://youtu.be/lg-KsjrlORo)

### OTHER ###


### STEPS TO VERIFY ###

1. Clone repo
2. Run `npm run test`
3. Write some NEW bad code (go ahead, you won't hurt anything)
4. Try and commit it

In step `#4` you may actually be able to commit it, but go look at your code!

![Magic!](https://media1.tenor.com/images/9678f9af8d211fdcd77088948ed92bbf/tenor.gif?itemid=11884906)

Write some more bad code, and try some commands out:

1. `npm run lint`
2. `npm run format`

The goal here is to make sure code is so freaking clean you could eat off of it!

![That was so clean!](https://media.tenor.com/images/e1ed31fbd548753d9494239184adf961/tenor.gif)

### DOCUMENTATION ###
Nope, it all just works.

![](https://media1.tenor.com/images/3274754b2961619d4158df4a45b97dd1/tenor.gif?itemid=15286450)

### CODE REVIEW TODO

- [ ] Start off with a fresh clone for best results
- [ ] I autoformatted all the files with `eslint` and `stylelint`, it shouldn't have broke anything, but can you make sure 🤞🏻
- [ ] Test the results of `npm run build` (can we use this at the end of the day) — looked okay to me installing it on a mock project
